extensions/fetch/lib.rs

```diff
@@ -38,24 +41,54 @@ use reqwest::Body;
 use reqwest::Client;
 use reqwest::Method;
 use reqwest::Response;
+use rustls::ClientConfig;
+use rustls::StoresClientSessions;
+use rustls_native_certs::load_native_certs;
 use serde::Deserialize;
 use serde::Serialize;
 use std::borrow::Cow;
 use std::cell::RefCell;
+use std::collections::HashMap;
 use std::convert::From;
 use std::fs::File;
 use std::io::Read;
 use std::path::Path;
 use std::path::PathBuf;
 use std::pin::Pin;
 use std::rc::Rc;
+use std::sync::Arc;
 use tokio::io::AsyncReadExt;
 use tokio::sync::mpsc;
 use tokio_stream::wrappers::ReceiverStream;
 use tokio_util::io::StreamReader;
 
 pub use reqwest; // Re-export reqwest
 
+lazy_static::lazy_static! {
+  static ref CLIENT_SESSION_MEMORY_CACHE: Arc<ClientSessionMemoryCache> =
+    Arc::new(ClientSessionMemoryCache::default());
+}
+
+#[derive(Default)]
+struct ClientSessionMemoryCache(Mutex<HashMap<Vec<u8>, Vec<u8>>>);
+
+impl StoresClientSessions for ClientSessionMemoryCache {
+  fn get(&self, key: &[u8]) -> Option<Vec<u8>> {
+    self.0.lock().get(key).cloned()
+  }
+
+  fn put(&self, key: Vec<u8>, value: Vec<u8>) -> bool {
+    let mut sessions = self.0.lock();
+    // TODO(bnoordhuis) Evict sessions LRU-style instead of arbitrarily.
+    while sessions.len() >= 1024 {
+      let key = sessions.keys().next().unwrap().clone();
+      sessions.remove(&key);
+    }
+    sessions.insert(key, value);
+    true
+  }
+}
```

Comment:
This is a duplicate of the cache from the `net` extension:
https://github.com/denoland/deno/blob/main/extensions/net/ops_tls.rs#L83

Also the websocket extension isn't using a cache at all, now that I look at it:
https://github.com/denoland/deno/blob/main/extensions/websocket/lib.rs#L225

Can I push this cache struct up somehwere higher in deno, so they can all reference it?

Reply:
I just duplicated it for now, there are two other spots where it should also either link to a shared function or duplicate this code again.

I'm not quite familiar with the code base enough to know what the right way to elevate the structs into a place where multiple extensions can access it. Perhaps a follow up PR? I don't believe it would be a regression to push this off as they are currently not using the cache either.

Reply:
I moved this into a shared location in the `deno_core` module so the code isn't duplicated and all clients share the same cache.